azurelinuxagent/ga/collect_telemetry_events.py

```diff
@@ -0,0 +1,585 @@
+            raise InvalidExtensionEventError(
+                key_err_msg.format(InvalidExtensionEventError.MissingKeyError, ExtensionEventSchema.Message))
+
+        if not event[message_key]:
+            raise InvalidExtensionEventError(
+                "{0}: {1} should not be empty".format(InvalidExtensionEventError.EmptyMessageError,
+                                                     ExtensionEventSchema.Message))
+
+        for required_key in self._EXTENSION_EVENT_REQUIRED_FIELDS:
+            # If all required keys not in event then raise
+            if required_key not in event:
+                raise InvalidExtensionEventError(
+                    key_err_msg.format(InvalidExtensionEventError.MissingKeyError, required_key))
+
+            # If the event_size > _EXTENSION_EVENT_MAX_SIZE=6k, then raise
+            if event_size > self._EXTENSION_EVENT_MAX_SIZE:
+                raise InvalidExtensionEventError(
+                    "{0}: max event size allowed: {1}".format(InvalidExtensionEventError.OversizeEventError,
+                                                              self._EXTENSION_EVENT_MAX_SIZE))
+
+            event_size += len(event[required_key])
+
+        return event
+
+    @staticmethod
+    def _replace_or_add_param_in_event(event, replace_or_add_params):
+        for param in event.parameters:
+            if param.name in replace_or_add_params:
+                param.value = replace_or_add_params.pop(param.name)
+
+        if not replace_or_add_params:
+            # All values replaced, return
+            return
+
+        # Add the remaining params to the event
+        for param_name in replace_or_add_params:
+            event.parameters.append(TelemetryEventParam(param_name, replace_or_add_params[param_name]))
+
+
+class _CollectAndEnqueueEventsPeriodicOperation(PeriodicOperation):
+    """
+    Periodic operation to collect and send telemetry events located in the events folder
+    """
+
+    _EVENT_COLLECTION_PERIOD = datetime.timedelta(minutes=1)
+
+    def __init__(self, send_telemetry_events_handler):
+        super(_CollectAndEnqueueEventsPeriodicOperation, self).__init__(
+            name="collect_and_enqueue_events",
+            operation=self._collect_and_enqueue_events,
+            period=_CollectAndEnqueueEventsPeriodicOperation._EVENT_COLLECTION_PERIOD)
+        self._send_telemetry_events_handler = send_telemetry_events_handler
+
+    def _collect_and_enqueue_events(self):
+        """
+        Periodically send any events located in the events folder
+        """
+        try:
+            if self._send_telemetry_events_handler.stopped():
+                logger.warn("{0} service is not running, skipping iteration.".format(
+                    self._send_telemetry_events_handler.get_thread_name()))
+                return
+            self.process_events()
+        except Exception as error:
+            err_msg = "Failure in collecting Agent events: {0}".format(ustr(error))
```

Comment:
These are not just agent events, right? While the extensions are not using the new pipeline, this will contain all events.

Reply:
Yes true but they're going to be transitioned to the new pipeline soon. It would make the message too complicated if I wrote something like "Failure in Agent/legacy extension events using old pipeline" or something. Keeping it simple basically :)

Reply:
Sure, but the message is misleading until ETP is enabled *and* until all extensions move to it. I'm concerned because this will be flagged as ERROR in the agent code and support/customers unfamiliar with the implementation might get the wrong idea from this message (like, what's happening to my extension's events?).

Today, this is the only avenue for events collection.

"Failure in collecting events using the legacy telemetry pipeline" captures the gist of it.

Reply:
> legacy telemetry pipeline

This is an internal concept though, like for the agent telemetry its not legacy, its current. Its only legacy for extensions events still using this pipeline. 

To tackle this exact behavior, I also plan on changing the location of our `events` directory to ensure these legacy extensions are forced to move. But we cant do that without proper communication. I can just rename it to `Failure in collecting telemetry events` if that sounds better to you? The ETP pipeline specifies `Extension Events` so that way we can be sure its not the new one,

Reply:
Sounds good to me! You are right about the legacy comment, it is a technical detail.

Reply:
Done